xarray/core/variable.py

```diff
@@ -1096,7 +1096,13 @@ def to_index(self):
         # basically free as pandas.Index objects are immutable
         assert self.ndim == 1
         index = self._data_cached().array
-        if not isinstance(index, pd.MultiIndex):
+        if isinstance(index, pd.MultiIndex):
+            # set default names for multi-index unnamed levels so that
+            # we can safely rename dimension / coordinate later
+            valid_level_names = [name or '{}_level_{}'.format(self.name, i)
```

Comment:
Yes it makes sense and it is more compact. However, how do we handle cases where two or more multi-index dimensions both collapse to `level_0` after indexing? Maybe such cases are rather edge cases so we don't need to care too much about this? Also a dimension / coordinate renamed to `level_0` is not very informative. Maybe we can include the old dimension name when renaming in `_replace_indexes`? Or, if we recommend xarray users to set explicit names for multi-index levels, we can probably leave this as-is...

I let you decide.